src/crypto/algorithms/megolm.js

```diff
@@ -262,15 +270,162 @@ MegolmEncryption.prototype._prepareNewSession = async function() {
  *
  * @param {module:crypto/algorithms/megolm.OutboundSessionInfo} session
  *
+ * @param {object<userId, deviceId>} devicemap
+ *   mapping from userId to deviceId to {@link module:crypto~OlmSessionResult}
+ *
  * @param {object<string, module:crypto/deviceinfo[]>} devicesByUser
  *    map from userid to list of devices
  *
+ * @return {array<object<userid, deviceInfo>>}
+ */
+MegolmEncryption.prototype._splitUserDeviceMap = function(
+    session, devicemap, devicesByUser,
+) {
+    const maxToDeviceMessagesPerRequest = 20;
+
+    // use an array where the slices of a content map gets stored
+    const contentMaps = [];
+    let maxContentMapId = 0; // start inserting in the first contentMap
+    let currentToDeviceId = 0;
+
+    const key = this._olmDevice.getOutboundGroupSessionKey(session.sessionId);
+
+    for (const userId in devicesByUser) {
+        if (!devicesByUser.hasOwnProperty(userId)) {
+            continue;
+        }
+        const devicesToShareWith = devicesByUser[userId];
+        const sessionResults = devicemap[userId];
+
+        for (let i = 0; i < devicesToShareWith.length; i++) {
+            const deviceInfo = devicesToShareWith[i];
+            const deviceId = deviceInfo.deviceId;
+
+            const sessionResult = sessionResults[deviceId];
+            if (!sessionResult.sessionId) {
+                // no session with this device, probably because there
+                // were no one-time keys.
+                //
+                // we could send them a to_device message anyway, as a
+                // signal that they have missed out on the key sharing
+                // message because of the lack of keys, but there's not
+                // much point in that really; it will mostly serve to clog
+                // up to_device inboxes.
+
+                // mark this device as "handled" because we don't want to try
+                // to claim a one-time-key for dead devices on every message.
+                session.markSharedWithDevice(userId, deviceId, key.chain_index);
+
+                // ensureOlmSessionsForUsers has already done the logging,
+                // so just skip it.
+                continue;
+            }
+
+            console.log(
+                "share keys with device " + userId + ":" + deviceId,
+            );
+
+            if (currentToDeviceId > maxToDeviceMessagesPerRequest) {
+                // the first slice is filled up. Start inserting into the next slice
+                currentToDeviceId = 0;
+                maxContentMapId++;
+            }
+            if (!contentMaps[maxContentMapId]) {
+                contentMaps[maxContentMapId] = [{
```

Comment:
can you make this `contentMaps[maxContentMapId] = []` and then make the `push` unconditional, to reduce duplication?